src/compiler/checker.ts

```diff
@@ -22630,7 +22633,153 @@ namespace ts {
+
+                if (propertyTypeArray.some(t =>
+                    t.reason === DeepPropertyUnreachableReason.OnlyNeverOnPath || t.reason === DeepPropertyUnreachableReason.OnlyNullableOnPath
+                    && rootNullable.every(n => n.id !== t.tyepeId)
+                )) {
+                    return true;
+                }
+
+                if (!(computedType.flags & TypeFlags.Union)) {
+                    // Whether we fall back to the declared type depends entirely on how
+                    // the type could be subsequently narrowed.
+                    /**
+                     * Condition 1: computed type is not union any more.
+                     *  // omit the defination, which is easy to be infered through code and intention
+                     *  function foo(x: A | B): any {
+                     *       x;  // A | B
+                     *       if ( 'A' === x.type ) { // first trigger
+                     *           return x;  // A
+                     *       }
+                     *       x;  // B
+                     *       if ('B' === x.type) {  // second trigger
+                     *           return x;  // B
+                     *       }
+                     *       x;  // In the second time narrowing, the computed type would only have one(not Union), but should still be triggered to narrow type to never.
+                     *   }
+                     */
+                    // use declaredType is a quick way to resolve issue above -- when computed type is not union, the declared type is not chagned.
+                    // But it also bring some other issue, like #39114 and #39110. As long as the declared type is not proper, issues come.
+
+                    // and only declared union is too loose to judge it could be narrowed.
+                    /**
+                     * Condition 2: Index type with undefined/null
+                     *   function ffff1() {
+                     *       let a: Array<number>|undefined;
+                     *       a = [1];
+                     *       if(a){
+                     *           if (a[2] !== undefined) {
+                     *               a[1];
+                     *           } else {
+                     *               a[2];  // declared type is union, but here should not be narrowed to never.
+                     *           }
+                     *       }
+                     *   }
+                     */
+                    if (!(declaredType.flags & TypeFlags.Union)) {
+                        return false;
+                    }
+                    //, the types expect primitive should have more than one.
+                    if ((<UnionType>declaredType).types.filter(t => !(t.flags & TypeFlags.Primitive)).length < 2) {
+                        return false;
+                    }

```

Comment:
I don’t understand this; can you explain?

Reply:
This is a optimisation. But it seems wired and useless when I revisit it today.
Origional purpose is :
```
interface UP{ foo:string };
type U = UP|string;
```
`U` could not be narrowed as discriminant, because primitive type could not has subtype. Nullable type was handled as special condition before, so here is correct.
But today, this optimisation is done in the core function `getPropertyTypesFromTypeAccordingToExpression`, so this is absolutely a useless burdon.